### PR TITLE
Pull title content from subsquare 

### PR DESCRIPTION
**Pull request by Lt-Kartik-9PARA-SF**
Show Title and Content from Subsquare when we don't have it

### CODE REVIEW

File: pages/api/v1/posts/subsqaure/subsquare-content.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+export const getSubSquareContentAndTitle = async (proposalType: string | string[], network: string | string[] | undefined, id: string | string[] | number |undefined) => {
+	//console.log('params', proposalType , network , id);
+	try {
+		if( typeof proposalType !== 'string' ){
+			throw apiErrorWithStatusCode('can not send String[] in Proposal type', 400);
+			return;
+		}
+		const url = urlMapper[String(proposalType)]?.(id, network);
+		const data = await (await fetch(url)).json();
```
[the remark this answers is not in the document]
Comment (alphainfinitus) [reply pages/api/v1/posts/subsqaure/subsquare-content.ts:36]: False Alarm : isValidNetwork() check implemented

File: src/components/Gov2Home/Gov2LatestActivity/AllGov2PostsTable.tsx [hunk 2]
```diff
@@ -118,7 +118,6 @@ const AllGov2PostsTable: FC<IAllGov2PostsTableProps> = ({ posts, error }) => {
 				let title = post.title || post.method || noTitle;
 				title = title.length > 80 ? `${title.substring(0, Math.min(80, title.length))}...`  : title.substring(0, Math.min(80, title.length));
 				const subTitle = !title && post.method;
-
```
[the remark this answers is not in the document]
Comment (Lt-Kartik-9PARA-SF) [reply src/components/Gov2Home/Gov2LatestActivity/AllGov2PostsTable.tsx:121]: will fix

File: pages/root/index.tsx [hunk 3]
```diff
@@ -108,6 +108,7 @@ const Root: FC<IRootProps> = (props) => {
 	if (error) return <ErrorState errorMessage={error} />;
 
 	if (!posts || Object.keys(posts).length === 0) return null;
+	console.log('posts',posts);
```
[the remark this answers is not in the document]
Comment (Lt-Kartik-9PARA-SF) [reply pages/root/index.tsx:111]: will fix

File: pages/referenda/[id].tsx [hunk 4]
```diff
@@ -80,7 +80,6 @@ const ReferendaPost: FC<IReferendaPostProps> = (props) => {
 				trackName = key;
 			}
 		}
-
```
[the remark this answers is not in the document]
Comment (Lt-Kartik-9PARA-SF) [reply pages/referenda/[id].tsx:83]: will fix

File: pages/api/v1/posts/subsqaure/subsquare-content.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+		const url = new URL( urlMapper[String(proposalType)]?.(id, network));
+		const data = await (await fetch(url)).json();
+
+		let subsqTitle = data.title;
+
+		subsqTitle = String(data?.title)?.includes('Untitled') ? '' : data.title;
+
+		if(subsqTitle){
+			subsqTitle.includes('[Root] Referendum #') ? subsqTitle = subsqTitle.replace(/\[Root\] Referendum #\d+: /, '') : '';
```
[the remark this answers is not in the document]
Comment (Lt-Kartik-9PARA-SF) [reply pages/api/v1/posts/subsqaure/subsquare-content.ts:65]: sometimes subsquare is sending title like "[Root] Referendum #45 Title text" and we are already showing referendum number already in our UI so i am replacing "[Root] Referendum #45 Title text"  with  " " and showing only the title text

File: pages/api/v1/posts/subsqaure/subsquare-content.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+		const subsquareData = { content : data.content ,title:subsqTitle };
+		return subsquareData;
+	} catch (error) {
+		console.log(`Error ${error}`);
+		return { content: '',title: '' };
+	}
+};
+
+const handler: NextApiHandler<{ data: any } | { error: string }> = async (req, res) => {
```
[the remark this answers is not in the document]
Comment (Lt-Kartik-9PARA-SF) [reply pages/api/v1/posts/subsqaure/subsquare-content.ts:76]: will fix

File: pages/api/v1/posts/subsqaure/subsquare-content.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+		subsqTitle = String(data?.title)?.includes('Untitled') ? '' : data.title;
+
+		if(subsqTitle){
+			subsqTitle.includes('[Root] Referendum #') ? subsqTitle = subsqTitle.replace(/\[Root\] Referendum #\d+: /, '') : '';
+		}
+
+		const subsquareData = { content : data.content ,title:subsqTitle };
+		return subsquareData;
+	} catch (error) {
+		console.log(`Error ${error}`);
```
[the remark this answers is not in the document]
Comment (Lt-Kartik-9PARA-SF) [reply pages/api/v1/posts/subsqaure/subsquare-content.ts:71]: will fix

File: pages/api/v1/posts/subsqaure/subsquare-content.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
+// Copyright 2019-2025 @polkassembly/polkassembly authors & contributors
+// This software may be modified and distributed under the terms
+// of the Apache-2.0 license. See the LICENSE file for details.
+import { NextApiHandler } from 'next';
+import withErrorHandling from '~src/api-middlewares/withErrorHandling';
+import { isValidNetwork } from '~src/api-utils';
+import { ProposalType } from '~src/global/proposalType';
+import apiErrorWithStatusCode from '~src/util/apiErrorWithStatusCode';
+
+const urlMapper: any = {
```
[the remark this answers is not in the document]
Comment (Lt-Kartik-9PARA-SF) [reply pages/api/v1/posts/subsqaure/subsquare-content.ts:10]: will fix

File: pages/api/v1/listing/on-chain-posts.ts [hunk 9]
```diff
@@ -544,6 +552,10 @@ export async function getOnChainPosts(params: IGetOnChainPostsParams): Promise<I
 						}
 					}
 
+					let subsquareTitle = '';
+					await getSubSquareContentAndTitle(strProposalType,network,postId).then((response) => {
```
[the remark this answers is not in the document]
Comment (Lt-Kartik-9PARA-SF) [reply pages/api/v1/listing/on-chain-posts.ts:556]: will fix

File: pages/api/v1/listing/on-chain-posts.ts [hunk 10]
```diff
@@ -515,6 +517,12 @@ export async function getOnChainPosts(params: IGetOnChainPostsParams): Promise<I
 					if (postDoc && postDoc.exists) {
 						const data = postDoc.data();
 						if (data) {
+							let subsquareTitle = '';
```
[the remark this answers is not in the document]
Comment (Lt-Kartik-9PARA-SF) [reply pages/api/v1/listing/on-chain-posts.ts:520]: will fix

File: pages/api/v1/posts/on-chain-post.ts [hunk 11]
```diff
@@ -902,6 +917,13 @@ export async function getOnChainPost(params: IGetOnChainPostParams) : Promise<IA
...
 				post.post_link = post_link;
 			}
+			else{
+				await getSubSquareContentAndTitle(proposalType,network,numPostId).then((response) => {
+					subsquareTitle = response?.title;subsquareContent = response?.content;
+				});
+				post.content = subsquareContent;
+				post.title = subsquareTitle;
+			}
 		}
```
[the remark this answers is not in the document]
Comment (Lt-Kartik-9PARA-SF) [reply pages/api/v1/posts/on-chain-post.ts:928]: will fix

File: pages/api/v1/latest-activity/all-posts.ts [hunk 12]
```diff
@@ -174,11 +175,24 @@ export async function getLatestActivityAllPosts(params: IGetLatestActivityAllPos
 				};
 				if (postDoc && postDoc.exists) {
 					const data = postDoc?.data();
+					let subsquareTitle = '';
+					if(data?.title === '' || data?.title === method || data?.title === null){
+						await getSubSquareContentAndTitle(getFirestoreProposalType(type) as ProposalType,network,postId).then((response) => {
```
[the remark this answers is not in the document]
Comment (Lt-Kartik-9PARA-SF) [reply pages/api/v1/latest-activity/all-posts.ts:180]: will fix

File: pages/api/v1/latest-activity/on-chain-posts.ts [hunk 13]
```diff
@@ -102,6 +103,12 @@ export async function getLatestActivityOnChainPosts(params: IGetLatestActivityOn
 			if (postDoc && postDoc.exists) {
 				const data = postDoc?.data();
 				if (data) {
+					let subsquareTitle = '';
+					if(data?.title === '' || data?.title === method || data.title === null){
+						await getSubSquareContentAndTitle(strProposalType as ProposalType,network,postId).then((response) => {
```
[the remark this answers is not in the document]
Comment (Lt-Kartik-9PARA-SF) [reply pages/api/v1/latest-activity/on-chain-posts.ts:108]: will fix

File: pages/api/v1/posts/on-chain-post.ts [hunk 14]
```diff
@@ -856,15 +857,14 @@ export async function getOnChainPost(params: IGetOnChainPostParams) : Promise<IA
 			}catch(e){
 				data=undefined;
 			}
+
 			// Populate firestore post data into the post object
 			if (data && post) {
 				post.topic = getTopicFromFirestoreData(data, strProposalType);
-				post.content = data.content;
```
[the remark this answers is not in the document]
Comment (Lt-Kartik-9PARA-SF) [reply pages/api/v1/posts/on-chain-post.ts]: added back

File: pages/api/v1/posts/on-chain-post.ts [hunk 15]
```diff
@@ -856,15 +857,14 @@ export async function getOnChainPost(params: IGetOnChainPostParams) : Promise<IA
 			}catch(e){
 				data=undefined;
 			}
+
```
[the remark this answers is not in the document]
Comment (Lt-Kartik-9PARA-SF) [reply pages/api/v1/posts/on-chain-post.ts]: removed

File: pages/api/v1/posts/subsqaure/subsquare-content.ts [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+		return { content: '',title: '' };
+	}
+};
+
+const handler: NextApiHandler<{ data: ({ content: any|string , title: string|any })| undefined} | { error: string } > = async (req, res) => {
+	const { proposalType, id } = req.query;
+	const network = String(req.headers['x-network']);
+
+	if (!network || !isValidNetwork(network)) res.status(400).json({ error: 'Invalid network in request header' });
+	if (!id || Array.isArray(id) ) res.status(400).json({ error: 'id missing in request' });
```
[the remark this answers is not in the document]
Comment (Lt-Kartik-9PARA-SF) [reply pages/api/v1/posts/subsqaure/subsquare-content.ts]: removed

File: pages/api/v1/posts/subsqaure/subsquare-content.ts [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+		if(!proposalType){
+			throw apiErrorWithStatusCode('Proposal type missing ', 400);
+			return;
+		}
+		if( typeof proposalType !== 'string' ){
+			throw apiErrorWithStatusCode('can not send String[] in Proposal type', 400);
+			return;
+		}
+
+		if(!id || Array.isArray(id)){
```
[the remark this answers is not in the document]
Comment (Lt-Kartik-9PARA-SF) [reply pages/api/v1/posts/subsqaure/subsquare-content.ts]: removed

File: pages/api/v1/posts/subsqaure/subsquare-content.ts [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+			throw apiErrorWithStatusCode('id is not present', 400);
+			return;
+		}
+		if(proposalType === ProposalType.TIPS && typeof id !== 'string'){
+			throw apiErrorWithStatusCode('type of id should be string', 400);
+			return;
+		}
+		if( proposalType !== ProposalType.TIPS && Number(id) < 0){
+			throw apiErrorWithStatusCode('id can not be negative', 400);
+			return;
```
[the remark this answers is not in the document]
Comment (Lt-Kartik-9PARA-SF) [reply pages/api/v1/posts/subsqaure/subsquare-content.ts]: removed

**Pull request merged**